command.c

```diff
@@ -1,49 +1,212 @@
-#include <stdio.h>
-#include <evbuffsock.h>
+#include "nsq.h"
 
-const static char * NEW_LINE = "\n";
-const static int MAX_BUF_SIZE = 128;
+#define NEW_LINE     "\n"
+#define NEW_SPACE    " "
+#define MAX_BUF_SIZE 128
+#define MIN_BUF_SIZE 64
+#define BUF_DELTER   2
 
-void nsq_subscribe(struct Buffer *buf, const char *topic, const char *channel)
+void *nsq_buf_malloc(size_t buf_size, size_t n, size_t l)
 {
-    char b[MAX_BUF_SIZE];
-    size_t n;
+    if (buf_size - n >= MIN_BUF_SIZE || buf_size - l >= MIN_BUF_SIZE) {
+        return NULL;
+    }
+    
+    void *buf = NULL;
+    while (1) {
+        if (buf_size - n < MIN_BUF_SIZE || buf_size - l < MIN_BUF_SIZE) {
+            buf_size *= BUF_DELTER;
+            continue;
+        }
+        buf = malloc(buf_size * sizeof(char));
+        assert(NULL != buf);
+        break;
+    }
+    
+    return buf;
+}
+
+void nsq_buffer_add(nsqBuf *buf, const char *name, const nsqCmdParams params[], size_t psize, const char *body, const size_t body_length)
+{
+    size_t buf_size = MAX_BUF_SIZE;
+    char *b = malloc(buf_size * sizeof(char));
+    char *nb = NULL;
+    assert(NULL != b);
+    size_t n = 0;
+    size_t l = 0;
+
+    l = strlen(name);
+    memcpy(b, name, l);
+    n += l;
+
+    if (NULL != params) {
+        for (int i = 0; i < psize; i++) {
+            memcpy(b+n, NEW_SPACE, 1);
+            n += 1;
 
-    n = sprintf(b, "SUB %s %s%s", topic, channel, NEW_LINE);
+            switch (params[i].t) {
+                case NSQ_PARAM_TYPE_INT:
+                    l = sprintf(b+n, "%d", *((int *)params[i].v));
+                    break;
+                case NSQ_PARAM_TYPE_CHAR:
+                    nb = nsq_buf_malloc(buf_size, n, strlen(params[i].v));
+                    if (NULL != nb) {
+                        memcpy(nb, b, n);
+                        free(b);
+                        b = nb;
+                    }
+                    l = strlen((char *)params[i].v);
+                    memcpy(b+n, (char *)params[i].v, l);
+                    break;
+            }
+            n += l;
+        }
+    }
+    memcpy(b+n, NEW_LINE, 1);
+    n += 1;
+
+    if (NULL != body) {
+        //convert to big endian
+        uint32_t v = (uint32_t)body_length;
+        uint32_t vv = (v & 0x000000ff) << 24 | (v & 0x0000ff00) << 16 | (v & 0x00ff0000) << 8 | (v & 0xff000000);
```

Comment:
this is equivalent to `htonl(body_length)`

Reply:
Thanks, I knows this, I just want to make it more unstandstandable. OK, I will change.